server-spi-private/src/main/java/org/keycloak/models/locking/GlobalLockProvider.java

```diff
@@ -17,57 +17,73 @@
+     * <p/>
      * If there is another global lock with the same identifier ({@code lockName}) already acquired, this method waits
      * until the lock is released, however, not more than {@code timeToWaitForLock} duration. If the lock is not
-     * acquired after {@code timeToWaitForLock} duration, the method throws {@link LockAcquiringTimeoutException}
-     * <p />
-     * Releasing of the lock is done using instance of {@link GlobalLock} returned by this method.
+     * acquired after {@code timeToWaitForLock} duration, the method throws {@link LockAcquiringTimeoutException}.
+     * <p/>
+     * When the execution of the {@code task} finishes, the acquired lock must be released regardless of the result.
+     * <p/>
+     * To make sure acquiring/releasing the lock is visible to all Keycloak nodes it may be needed to run the code that
+     * acquires/releases the lock in a separate transactions. This means together the method can use 3 separate
+     * transactions, for example:
+     * <pre>
+     *     try {
+     *         KeycloakModelUtils.runJobInTransaction(factory,
+     *                                innerSession -> /* run code that acquires the lock *\/)
+     *
+     *         KeycloakModelUtils.runJobInTransactionWithResult(factory, task)
+     *     } finally {
+     *         KeycloakModelUtils.runJobInTransaction(factory,
+     *                                innerSession -> /* run code that releases the lock *\/)
+     *     }
+     * </pre>
      *
      * @param lockName Identifier used for acquiring lock. Can be any non-null string.
+     * @param task The task that will be executed under the acquired lock
+     * @param <V> Type of object returned by the {@code task}
      * @param timeToWaitForLock Duration this method waits until it gives up acquiring the lock. If {@code null},
      *                          each implementation should provide some default duration, for example using
      *                          configuration option.
-     * @return Instance of {@link GlobalLock} representing successfully acquired global lock.
+     * @return Value returned by the {@code task}
      *
      * @throws LockAcquiringTimeoutException When the method waits for {@code timeToWaitForLock} duration and the lock is still
-     *                          not available to acquire.
-     * @throws NullPointerException When {@code lockName} is {@code null}.
+     *                                       not available to acquire.
+     * @throws NullPointerException          When {@code lockName} is {@code null}.
      */
-    GlobalLock acquire(String lockName, Duration timeToWaitForLock) throws LockAcquiringTimeoutException;
+    <V> V withLock(String lockName, Duration timeToWaitForLock, KeycloakSessionTaskWithResult<V> task) throws LockAcquiringTimeoutException;
 
     /**
      * Releases all locks acquired by this GlobalLockProvider.
```

Comment:
In this JavaDoc, please also change "This method must unlock all existing..." to "This method unlocks all existing..." to take the view of a caller.